src/Wallet.ts

```diff
@@ -0,0 +1,252 @@
+  };
+
+  private fetchAddressCollectionStats = async (
+    node: bip32.BIP32Interface,
+  ): Promise<AddressCollectionStats> => {
+    let gap = 0;
+    let addressIndex = 0;
+    let balance = 0;
+    let pendingBalance = 0;
+    let addresses: AddressMetadata[] = [];
+    let addressesWithUtxo: AddressMetadata[] = [];
+    let nextUnusedAddress: AddressMetadata | null = null;
+    while (gap < this.gapLimit) {
+      const childNode = node.derive(addressIndex);
+      const { address } = bitcoinjs.payments.p2wpkh({
+        pubkey: childNode.publicKey,
+      });
+
+      const addressStats = await this.blockchainService.fetchAddressStats(
+        address!,
+      );
+
+      const { chain_stats, mempool_stats } = addressStats;
+
+      const isUsedAddress =
+        chain_stats.funded_txo_count > 0 ||
+        chain_stats.spent_txo_count > 0 ||
+        mempool_stats.funded_txo_count > 0 ||
+        mempool_stats.spent_txo_count > 0;
+
+      const addressBalance =
+        chain_stats.funded_txo_sum - chain_stats.spent_txo_sum;
+
+      balance += addressBalance;
+      pendingBalance +=
+        mempool_stats.funded_txo_sum - mempool_stats.spent_txo_sum;
+
+      const addressMetadata = {
+        address: address!,
+        derivationPath: `${node.index}/${addressIndex}`,
+        publicKey: childNode.publicKey,
+      };
+
+      addresses.push(addressMetadata);
+
+      if (addressBalance > 0) {
+        addressesWithUtxo.push(addressMetadata);
+      }
+
+      if (!isUsedAddress && !nextUnusedAddress) {
+        nextUnusedAddress = addressMetadata;
+      }
+
+      if (!isUsedAddress) {
+        gap++;
+      } else {
+        gap = 0;
+      }
+
+      addressIndex++;
+    }
+
+    return {
+      balance,
+      pendingBalance,
+      addresses,
+      addressesWithUtxo,
+      nextUnusedAddress: nextUnusedAddress!,
+    };
+  };
+
+  fetchWalletStats = async (masterPublicKey: string): Promise<WalletStats> => {
+    const hdNode = bip32.fromBase58(masterPublicKey);
+    const [addressesData, changeAddressesData] = await Promise.all([
+      this.fetchAddressCollectionStats(hdNode.derive(0)),
+      this.fetchAddressCollectionStats(hdNode.derive(1)),
+    ]);
+
+    const balance = addressesData.balance + changeAddressesData.balance;
+    const pendingBalance =
+      addressesData.pendingBalance + changeAddressesData.pendingBalance;
+    const nextUnusedAddress = addressesData.nextUnusedAddress;
+    const nextUnusedChangeAddress = changeAddressesData.nextUnusedAddress;
+    const addresses = addressesData.addresses;
+    const changeAddresses = changeAddressesData.addresses;
+    const addressesWithUtxo = addressesData.addressesWithUtxo.concat(
+      changeAddressesData.addressesWithUtxo,
+    );
+
+    return {
+      balance,
+      pendingBalance,
+      nextUnusedAddress,
+      nextUnusedChangeAddress,
+      addresses,
+      changeAddresses,
+      addressesWithUtxo,
+    };
+  };
+
+  fetchUtxos = async (
```

Comment:
Using the list of UTXO addresses retrieved from `fetchWalletStats`, the developer can use this method to retrieve a list of UTXOs belonging to each one of those addresses.

This method augments the retrieved UTXO object with `address` and `derivationPath`. This makes these values conveniently available during the transaction creation process.